cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs

```diff
@@ -321,22 +325,48 @@ votingScriptWitnessSingleton votingProcedures (Just scriptWitness) =
   let voter = fromJust $ getVotingScriptCredentials votingProcedures
   in Map.singleton voter scriptWitness
 
+newtype TxGovDuplicateVotes era =
+  TxGovDuplicateVotes [L.GovActionId (L.EraCrypto (ShelleyLedgerEra era))]
+
+instance Error (TxGovDuplicateVotes era) where
+  prettyError (TxGovDuplicateVotes actionIds) =
+    "Trying to merge votes with similar action identifiers: " <> viaShow actionIds <>
+      ". This would cause ignoring some of the votes, so not proceeding."
+
 -- TODO: We fold twice, we can do it in a single fold
 convertToTxVotingProcedures
  :: [(VotingProcedures era, Maybe (ScriptWitness WitCtxStake era))]
- -> TxVotingProcedures BuildTx era
-convertToTxVotingProcedures votingProcedures =
-  let votingScriptWitnessMap = BuildTxWith
-                                 $ foldl (\acc next -> acc `Map.union` uncurry votingScriptWitnessSingleton next )
-                                   Map.empty
-                                   votingProcedures
-      allVotes = unVotingProcedures
-                   $ foldl
-                       (\acc next -> acc `unsafeMergeVotingProcedures` fst next)
-                       emptyVotingProcedures
-                       votingProcedures
-
-  in TxVotingProcedures allVotes votingScriptWitnessMap
+ -> Either (TxGovDuplicateVotes era) (TxVotingProcedures BuildTx era)
+convertToTxVotingProcedures votingProcedures = do
+  VotingProcedures procedure <- foldM f emptyVotingProcedures votingProcedures
+  pure $ TxVotingProcedures procedure (BuildTxWith votingScriptWitnessMap)
+  where
+    votingScriptWitnessMap = foldl (\acc next -> acc `Map.union` uncurry votingScriptWitnessSingleton next)
+                               Map.empty
+                               votingProcedures
+    f acc (procedure, _witness) = mergeVotingProcedures acc procedure
+
+mergeVotingProcedures :: ()
+  => VotingProcedures era
+  -> VotingProcedures era
+  -> Either (TxGovDuplicateVotes era) (VotingProcedures era) -- ^ Either an error message, or the merged voting procedures
+mergeVotingProcedures vpsa vpsb =
+  VotingProcedures . L.VotingProcedures <$> foldM mergeVotesOfOneVoter Map.empty allVoters
+  where
+    mapa = L.unVotingProcedures (unVotingProcedures vpsa)
+    mapb = L.unVotingProcedures (unVotingProcedures vpsb)
+    allVoters = Set.union (Map.keysSet mapa) (Map.keysSet mapb)
+    mergeVotesOfOneVoter acc voter =
+      Map.union acc <$> case (Map.lookup voter mapa, Map.lookup voter mapb) of
```

Comment:
We should be able to us [differenceWith](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Strict.html#v:differenceWith) here.